Libraries/libutil/Sources/DefaultProcessContext.cpp

```diff
@@ -0,0 +1,197 @@
+/**
+ Copyright (c) 2015-present, Facebook, Inc.
+ All rights reserved.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree. An additional grant
+ of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include <libutil/DefaultProcessContext.h>
+#include <libutil/FSUtil.h>
+
+#include <sstream>
+#include <unordered_set>
+#include <cstring>
+#include <cassert>
+
+#include <pwd.h>
+#include <grp.h>
+#include <unistd.h>
+#include <sys/select.h>
+#include <errno.h>
+
+#if defined(__APPLE__)
+#include <mach-o/dyld.h>
+#elif defined(__linux__)
+#include <linux/limits.h>
+#if __GLIBC__ >= 2 && __GLIBC_MINOR__ >= 16
+#include <sys/auxv.h>
+#endif
+#endif
+
+extern "C" char **environ;
+
+using libutil::DefaultProcessContext;
+using libutil::FSUtil;
+
+DefaultProcessContext::
+DefaultProcessContext() :
+    ProcessContext()
+{
+}
+
+DefaultProcessContext::
+~DefaultProcessContext()
+{
+}
+
+std::string DefaultProcessContext::
+currentDirectory() const
+{
+    char path[PATH_MAX + 1];
+    if (::getcwd(path, sizeof(path)) == nullptr) {
+        path[0] = '\0';
+    }
+    return path;
+}
+
+#if defined(__linux__)
+static char initialWorkingDirectory[PATH_MAX] = { 0 };
+__attribute__((constructor))
+static void InitializeInitialWorkingDirectory()
+{
+    if (getcwd(initialWorkingDirectory, sizeof(initialWorkingDirectory)) == NULL) {
+        abort();
+    }
+}
+
+#if !(__GLIBC__ >= 2 && __GLIBC_MINOR__ >= 16)
+static char initialExecutablePath[PATH_MAX] = { 0 };
+__attribute__((constructor))
+static void InitialExecutablePathInitialize(int argc, char **argv)
+{
+    strncpy(initialExecutablePath, argv[0], sizeof(initialExecutablePath));
+}
+#endif
+#endif
+
+std::string DefaultProcessContext::
+executablePath() const
+{
+#if defined(__APPLE__)
+    uint32_t size = 0;
+    if (_NSGetExecutablePath(NULL, &size) != -1) {
+        abort();
+    }
+
+    std::string buffer;
+    buffer.resize(size);
+    if (_NSGetExecutablePath(&buffer[0], &size) != 0) {
+        abort();
+    }
+
+    return FSUtil::NormalizePath(buffer);
+#elif defined(__linux__)
+#if __GLIBC__ >= 2 && __GLIBC_MINOR__ >= 16
```

Comment:
nit: move this define somewhere else, like #define IS_GLIBC_2_16, since you use it more than once.